docs/proposal/2022-12-23-rbac.md

```diff
@@ -0,0 +1,399 @@
+
+Created on 2022-12-23 by Paweł Kosiec ([@pkosiec](https://github.com/pkosiec))
+
+| Status     |
+| ---------- |
+| `PROPOSED` |
+
+## Overview
+
+Botkube exposes an ability to run arbitrary commands, especially those related to Kubernetes clusters. Currently, to execute any `kubectl` command, Botkube uses ClusterRole which comes with an installation. This ClusterRole is bound to a ServiceAccount used by a Botkube Pod. This raises security concerns.
+
+Also, while Botkube supports per-channel kubectl execution configuration, users should be allowed to set per-user or group mapping for more complex scenarios. While initially we could stick to channel-based grouping, it would be great to have an extensible mechanism to cover future use cases.
+
+<!-- toc -->
+
+- [Goal](#goal)
+- [Non-Goals](#non-goals)
+- [Assumptions](#assumptions)
+
+* [Proposal](#proposal)
+  - [User and group identification](#user-and-group-identification)
+    - [Implementation details](#implementation-details)
+  - [Configure mapping](#configure-mapping)
+  - [Executing commands with proper access rights](#executing-commands-with-proper-access-rights)
+  - [Executor plugin isolation](#executor-plugin-isolation)
+* [Reference](#reference)
+* [Alternatives](#alternatives)
+  - [Configure mapping](#configure-mapping-1)
+  - [Executing commands with proper access rights](#executing-commands-with-proper-access-rights-1)
+* [Consequences](#consequences)
+
+<!-- tocstop -->
+
+### Goal
+
+- Stop using ClusterRole assigned to the Botkube pod for executing commands.
+- Implement ability to configure command execution permissions for different user groups.
+- Add ability to map users from the communication platforms to Kubernetes permissions.
+- Ensure that the proposed approach can be used with external identity providers.
+- Support plugins.
+
+### Non-Goals
+
+- Changing RBAC approach for Sources.
+
+  Sources are not related to any specific user or group. They will still use ClusterRole permissions assigned to the Botkube Pod.
+
+### Assumptions
+
+- RBAC support is plugin-specific. Not all executors are based on Kubernetes API.
+
+## Proposal
+
+### User and group identification
+
+To identify users and groups, we have multiple options, which varies between communication platforms.
+Summarizing the options, the proposal suggests the following approaches:
+
+- User identification:
+  - User email
+  - Static value for a given configuration
+- Group identification:
+  - Channel name
+  - User groups (Slack, Mattermost), Roles (Discord). Not supported by MS Teams.
+  - Static value for a given configuration
+
+We can support just a few options initially, and later add more based on user feedback.
+
+To get user email, channel name or user group name, we would need to do additional API calls in bot logic. However, we can cache the fetched values.
+
+#### Implementation details
+
+Slack:
+
+- We need to add two new scopes to the Slack app: `users:read` and `users:read.email`
```

Comment:
I know that for interactive messages we have the username and user email already in the request, so we can extract it and use it for a given command execution flow.

Probably the same should be also for normal messages 🤔

Reply:
Please take a look at the PoC doc - for "normal" messages we don't have the email address, just user ID. Checked Slack, Discord, and Mattermost with verbose logging of incoming events, and for all of them this is the case. Teams docs also mention a need of additional query.